File: desktop/src/lib/types/library.ts
export type LibraryState =
    | "WATCHING"
    | "PLANNING"
    | "COMPLETED"
    | "REWATCHING"
    | "DROPPED"
    | "PAUSED";

export type LibraryListScope = "ALL" | LibraryState;

export type LibrarySortValue =
    | "A_Z"
    | "Z_A"
    | "POPULARITY_DESC"
    | "POPULARITY_ASC"
    | "USER_SCORE_DESC"
    | "USER_SCORE_ASC"
    | "RATING_DESC"
    | "RATING_ASC"
    | "PROGRESS_DESC"
    | "PROGRESS_ASC"
    | "OLDEST"
    | "NEWEST";

export type LibraryReleaseFilter = "ALL" | "FINISHED" | "RELEASING" | "NOT_YET_RELEASED";
export type KnownLibraryReleaseStatus = Exclude<LibraryReleaseFilter, "ALL">;

export interface LibrarySectionDef {
    value: LibraryState;
    label: string;
    dotClass: string;
}

export interface LibraryEntry {
    anilist_id: number;
    title: string;
    cover_image: string | null;
    banner_image: string | null;
    format: string | null;
    episode_total: number | null;
    score: number | null;
    genres: string[];
    anilist_status: string | null;
    season: string | null;
    season_year: number | null;
    popularity: number | null;
    average_score: number | null;
    start_date: string | null;
    end_date: string | null;
    rewatches: number;
    notes: string | null;
    status: LibraryState;
    progress: number;
    progress_percent: number;
    last_episode: number | null;
    last_watched_at: number | null;
    created_at: number;
    updated_at: number;
}

export interface LibraryFacets {
    genres: string[];
    release_statuses: string[];
    seasons: string[];
    years: number[];
}

File: core/crates/domain/src/library.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ShelfState {
    Watching,
    Planning,
    Completed,
    Rewatching,
    Dropped,
    Paused,
}

impl ShelfState {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Watching => "WATCHING",
            Self::Planning => "PLANNING",
            Self::Completed => "COMPLETED",
            Self::Rewatching => "REWATCHING",
            Self::Dropped => "DROPPED",
            Self::Paused => "PAUSED",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "WATCHING" => Some(Self::Watching),
            "PLANNING" => Some(Self::Planning),
            "COMPLETED" => Some(Self::Completed),
            "REWATCHING" => Some(Self::Rewatching),
            "DROPPED" => Some(Self::Dropped),
            "PAUSED" => Some(Self::Paused),
            _ => None,
        }
    }
}

impl Default for ShelfState {
    fn default() -> Self {
        Self::Planning
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ShelfEntry {
    pub anilist_id: u32,
    pub title: String,
    pub cover_image: Option<String>,
    pub banner_image: Option<String>,
    pub format: Option<String>,
    pub episode_total: Option<u32>,
    pub score: Option<u8>,
    #[serde(default)]
    pub genres: Vec<String>,
    pub anilist_status: Option<String>,
    pub season: Option<String>,
    pub season_year: Option<u32>,
    pub popularity: Option<u32>,
    pub average_score: Option<u8>,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
    #[serde(default)]
    pub rewatches: u32,
    pub notes: Option<String>,
    pub status: ShelfState,
    pub progress: u32,
    pub progress_percent: u8,
    pub last_episode: Option<u32>,
    pub last_watched_at: Option<i64>,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct EpisodeState {
    pub anilist_id: u32,
    pub episode: u32,
    pub percent: u8,
    pub watched_at: i64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WatchEvent {
    pub id: Option<i64>,
    pub anilist_id: u32,
    pub episode: u32,
    pub percent: u8,
    pub watched_at: i64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum SyncKind {
    Update,
    Progress,
    Remove,
}

impl SyncKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Update => "UPDATE",
            Self::Progress => "PROGRESS",
            Self::Remove => "REMOVE",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "UPDATE" => Some(Self::Update),
            "PROGRESS" => Some(Self::Progress),
            "REMOVE" => Some(Self::Remove),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SyncJob {
    pub id: Option<i64>,
    pub anilist_id: u32,
    pub kind: SyncKind,
    pub payload: String,
    pub attempts: u32,
    pub created_at: i64,
    pub sent_at: Option<i64>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LibraryFacets {
    #[serde(default)]
    pub genres: Vec<String>,
    #[serde(default)]
    pub release_statuses: Vec<String>,
    #[serde(default)]
    pub seasons: Vec<String>,
    #[serde(default)]
    pub years: Vec<u32>,
}

#[cfg(test)]
mod tests {
    use super::ShelfState;

    #[test]
    fn shelf_state_round_trip_includes_rewatching() {
        assert_eq!(
            ShelfState::parse("REWATCHING"),
            Some(ShelfState::Rewatching)
        );
        assert_eq!(ShelfState::Rewatching.as_str(), "REWATCHING");
    }
}